assembly/src/assembler/mod.rs

```diff
@@ -1,107 +1,308 @@
+///
+/// # Assembly Contexts
+///
+/// Using the instructions above, all of the code you provide will be compiled and cached using a
+/// single global context. That works fine if you are creating the assembler and discarding it after
+/// you've compiled your program. However, if you plan to compile multiple distinct programs, you
+/// will want to use [AssemblyContext]s and [Assembler::compile_in_context].
+///
+/// An [AssemblyContext] is essentially a way to isolate the program-specific elements of a
+/// compilation session in a separate cache, so that you avoid polluting the global cache with a
+/// bunch of objects from multiple programs, causing analysis to become more expensive. By isolating
+/// those in a context-specific cache, you have more fine control over how things are cached.
 ///
-/// The assembler can be instantiated in several ways using a "builder" pattern. Specifically:
-/// - If `with_kernel()` or `with_kernel_module()` methods are not used, the assembler will be
-///   instantiated with a default empty kernel. Programs compiled using such assembler
-///   cannot make calls to kernel procedures via `syscall` instruction.
-#[derive(Default)]
+/// More precisely, the [Assembler] has a global module graph and procedure cache, which it uses to
+/// perform analysis during compilation, and to avoid redundantly compiling the same procedures for
+/// every program. Any modules you add to the global context will be inherited by _all_ contexts.
+/// This is where you will cache the kernel, standard libraries, anything else that is quite common.
+///
+/// Each [AssemblyContext] has its own module graph and procedure cache, which contains only those
+/// modules which you add to it. When you call [Assembler::compile_in_context] with that context, it
+/// is merged with the global context, inter-procedural analysis is performed, and then the compiled
+/// objects are cached in the provided [AssemblyContext], allowing it to be used multiple times if
+/// desired.
+///
+/// <div class="warning">
+/// The context isolation described above is not currently how things are implemented, but I believe
+/// represent where we will want to ultimately take the `AssemblyContext` struct. The main obstacle
+/// right now is that we don't have a clear picture of how an `Assembler` will be used, so we want
+/// to make sure that we design the `Assembler` and `AssemblyContext` relationship in such a way
+/// that it plays well with the most common usage patterns.
+/// </div>
 pub struct Assembler {
-    kernel: Kernel,
-    module_provider: ModuleProvider,
-    proc_cache: RefCell<ProcedureCache>,
+    /// The global [ModuleGraph] for this assembler. All new [AssemblyContext]s inherit this graph
+    /// as a baseline.
+    module_graph: Box<ModuleGraph>,
+    /// The global procedure cache for this assembler.
+    procedure_cache: ProcedureCache,
+    /// Whether to treat warning diagnostics as errors
+    warnings_as_errors: bool,
```

Comment:
It seems a bit odd that we have `warnings_as_errors` both here and in `CompileOpts` - ideally, the user would just have one way of specifying these.

From what I can tell, we need this property in `CompileOpts` so that we can communicate this info to the parser. If this is correct, I think we could simplify this in two ways:

1. Make `warnings_as_errors` in compile options `pub(crate)`. This way, from the user's standpoint, these won't be visible but we could still set them from the assembler.
2. Remove `warnings_as_errors` from compile options and change the signature of `Compile::compile_with_options` to something like this:

```Rust
fn compile_with_options(self, options: Options, warnings_as_errors: bool) -> Result<Box<Module>, Report>;
```